src/PackageManagement/AzureAutomationTools.PackageManagement.psm1

```diff
@@ -767,27 +921,60 @@ function DeployModules {
+            $AutomationModule = $null
             
-                Write-Output "Updating existing module - done."
+            try {
+                $ExistingModule = Get-AzureRmAutomationModule @Params
+            }
+            catch [Microsoft.Azure.Commands.Automation.Common.ResourceNotFoundException] {
+                Write-Verbose "Cannot find module '$($Module.Name)' in automation account '$AutomationAccountName'."
+                $ExistingModule = $null
+            }
+
+            if ($null -eq $ExistingModule) {
+                Write-Output "Adding new module: $($Module.Name) $($Module.Version)"
+                $AutomationModule = New-AzureRmAutomationModule -ContentLink $ContentLink @Params
+            }
+            elseif ($ExistingModule.IsGlobal) {
+                Write-Output "Updating global module $($Module.Name) $($ExistingModule.Version) --> $($Module.Version)"
+                $AutomationModule = New-AzureRmAutomationModule -ContentLink $ContentLink @Params
+            }
+            elseif ($ExistingModule.Version -ne $Module.Version) {
+                Write-Output "Changing existing module: $($Module.Name) $($ExistingModule.Version) --> $($Module.Version)"
+                $AutomationModule = Set-AzureRmAutomationModule -ContentLinkUri $Module.Package -ContentLinkVersion $Module.Version @Params
             }
             else {
-                Write-Output "Adding new module ..."
-            
-                New-AzureRmAutomationModule -ContentLink $ContentLink @Params
+                Write-Output "Module up to date: $($Module.Name) $($Module.Version)"
+            }
+
+            if ($null -ne $AutomationModule) {
+                $PollCount = 0
+                # Wait by default (for dependency purposes). TODO: Add this to the module config
+                while($PollCount -lt $MaxModulePoll -and 
+                    $AutomationModule.ProvisioningState -ne 'Succeeded' -and
+                    $AutomationModule.ProvisioningState -ne 'Failed') {
+                        $PollCount++
+                        Start-Sleep -Seconds 10
+                        Write-Verbose "Polling for $($Module.Name) completion..."
+                        $AutomationModule = $AutomationModule | Get-AzureRmAutomationModule
+                }
+
+                if ($PollCount -eq $MaxModulePoll -and 
+                    $AutomationModule.ProvisioningState -ne 'Succeeded' -and
+                    $AutomationModule.ProvisioningState -ne 'Failed') {
+                    Write-Error -Message "Failed to upload module: $($Module.Name) - PollCount reached max limit."
+                    continue
+                }
 
-                Write-Output "Adding new module - done."
+                if ($AutomationModule.ProvisioningState -eq 'Failed') {
+                    Write-Error -Message "Failed to upload module: $($Module.Name) - ProvisioningState Failed"
+                    continue
```

Comment:
By default this should not continue on failure - it should throw. (this could be added later but we need to be fed through from **Publish**, so should be backlogged for now.